### PR TITLE
Lesson06

### PR DESCRIPTION
**Pull request by yarikkk13**

### CODE REVIEW

File: src/app/modules/courses/components/edit-course/edit-course.component.html [hunk 1]
```diff
@@ -1,24 +1,25 @@
+<app-breadcrumb></app-breadcrumb>
 <form class="form">
```
Comment (AndriyLeskiv) [inline src/app/modules/courses/components/edit-course/edit-course.component.html:2]: for first time remove form, it wil be in next module, use simple <input>

File: src/app/modules/courses/components/edit-course/edit-course.component.html [hunk 2]
```diff
@@ -1,24 +1,25 @@
...
   <mat-form-field class="full-width" appearance="outline">
     <mat-label>{{ title }}</mat-label>
-    <input matInput placeholder="Title">
+    <input matInput placeholder="Title" [value]="course?.title">
```
Comment (AndriyLeskiv) [inline src/app/modules/courses/components/edit-course/edit-course.component.html:5]: use ngModel two way binding

File: src/app/modules/courses/components/new-course/new-course.component.ts [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
 
 @Component({
   selector: 'app-new-course',
   templateUrl: './new-course.component.html',
   styleUrls: ['./new-course.component.css']
 })
+
 export class NewCourseComponent implements OnInit {
 
+  @Input()
```
Comment (AndriyLeskiv) [inline src/app/modules/courses/components/new-course/new-course.component.ts:13]: remove inpur, create empty object ICourse

File: src/app/modules/courses/components/breadcrumb/breadcrumb.component.html [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <nav aria-label="breadcrumb">
   <ol class="breadcrumb bg-white" style="padding-left: 4.5%">
-    <li class="breadcrumb-item active" aria-current="page">{{ courses }}</li>
+    <li class="breadcrumb-item active" aria-current="page">{{ courses }}/ {{ course?.id }}</li>
```
[the remark this answers is not in the document]
Comment (AndriyLeskiv) [reply src/app/modules/courses/components/breadcrumb/breadcrumb.component.html:3]: add course.title or courses value

File: src/app/services/courses.service.ts [hunk 5]
```diff
@@ -13,15 +13,15 @@ export class CoursesService {
...
+  public getCourseById(id: string): ICourse | undefined {
+    return courses.find(course => course.id == +id)
   }
 
   public createCourse(course: ICourse): void {
     courses.push(course)
   }
 
-  public updateCourse(id: string, course: ICourse): ICourse[] {
+  public updateCourse(id: number, course: ICourse): ICourse[] {
```
Comment (AndriyLeskiv) [inline src/app/services/courses.service.ts:24]: remove return, use forEach

File: src/app/services/courses.service.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
-import { Injectable } from '@angular/core';
+import {Injectable} from '@angular/core';
 
-import { courses } from "../mockArrays/mockCourses";
-import { ICourse } from "../models/Course";
+import {courses} from "../mockArrays/mockCourses";
```
Comment (AndriyLeskiv) [inline src/app/services/courses.service.ts:3]: save courses to value in constructor

File: src/app/app-routing.module.ts [hunk 7]
```diff
@@ -1,16 +1,20 @@
...
+
+import { HomeComponent } from "./components/home/home.component";
+import { NotFoundComponent } from "./components/not-found/not-found.component";
 
 const routes: Routes = [
-  {path: 'login', component: LoginComponent},
-  {path: 'add_course', component: NewCourseComponent},
+  {path: '', component: HomeComponent},
+  {path: 'login', loadChildren: () => import('./modules/login/login.module').then(m => m.LoginModule)},
+  {path: 'courses', loadChildren: () => import('./modules/courses/courses.module').then(m => m.CoursesModule)},
```
Comment (AndriyLeskiv) [inline src/app/app-routing.module.ts:10]: add canActivate guard https://metanit.com/web/angular2/7.7.php

File: src/app/modules/courses/components/edit-course/edit-course.component.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import {ICourse} from "../../../../models/Course";
+
+@Component({
+  selector: 'app-edit-course',
+  templateUrl: './edit-course.component.html',
+  styleUrls: ['./edit-course.component.css']
+})
+export class EditCourseComponent implements OnInit {
+
+  @Input()
```
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/modules/courses/components/edit-course/edit-course.component.ts:13]: done

File: src/app/modules/courses/components/breadcrumb/breadcrumb.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/modules/courses/components/breadcrumb/breadcrumb.component.html:3]: done

File: src/app/components/not-found/not-found.component.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+})
+export class NotFoundComponent implements OnInit {
+
+  public notFoundText = '404 error';
+
+  public text = "sorry, your page doesn't exists";
+
+  constructor() { }
+
+  ngOnInit(): void {
```
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/components/not-found/not-found.component.ts:16]: done

File: src/app/modules/courses/components/new-course/new-course.component.html [hunk 10]
```diff
@@ -0,0 +1,42 @@
+<app-breadcrumb></app-breadcrumb>
```
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/modules/courses/components/new-course/new-course.component.html:1]: done

File: src/app/modules/courses/components/edit-course/edit-course.component.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  constructor(private activatedRoute: ActivatedRoute,
+              private coursesService: CoursesService) {
+
+    this.activatedRoute.params
+      .subscribe(params => {
+        this.course = this.coursesService.getCourseById(params.id)
+      })
+  }
+
+  editCurrentCourse(id:number, course:ICourse) {
```
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/modules/courses/components/edit-course/edit-course.component.ts:36]: done

File: src/app/modules/courses/components/edit-course/edit-course.component.html (hunk 1 above)
Comment (yarikkk13) [reply src/app/modules/courses/components/edit-course/edit-course.component.html:2]: done

File: src/app/services/courses.service.ts (hunk 5 above)
Comment (yarikkk13) [reply src/app/services/courses.service.ts:24]: done

File: src/app/services/courses.service.ts (hunk 6 above)
Comment (yarikkk13) [reply src/app/services/courses.service.ts:3]: done

File: src/app/modules/courses/components/edit-course/edit-course.component.html [hunk 12]
```diff
@@ -33,7 +34,7 @@
   </mat-form-field>
 
   <div class="example-button-row">
-    <button mat-stroked-button>{{save}}</button>
-    <button mat-stroked-button>{{cancel}}</button>
+    <button mat-stroked-button routerLink="/courses">{{save}}</button>
```
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/modules/courses/components/edit-course/edit-course.component.html:37]: done

File: src/app/modules/courses/components/new-course/new-course.component.ts (hunk 3 above)
Comment (yarikkk13) [reply src/app/modules/courses/components/new-course/new-course.component.ts:13]: done

File: src/app/modules/courses/components/edit-course/edit-course.component.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+  public authors: string = 'Authors';
+  public save: string = 'Save';
+  public cancel: string = 'Cancel';
+
+
+  constructor(private activatedRoute: ActivatedRoute,
+              private coursesService: CoursesService) {
+
+    this.activatedRoute.params
+      .subscribe(params => {
```
[the remark this answers is not in the document]
Comment (yarikkk13) [reply src/app/modules/courses/components/edit-course/edit-course.component.ts:24]: done

File: src/app/app-routing.module.ts (hunk 7 above)
Comment (yarikkk13) [reply src/app/app-routing.module.ts:10]: done

File: src/app/services/courses.service.ts [hunk 14]
```diff
@@ -1,37 +1,44 @@
...
       }
-      return currentCourse
     })
   }
 
   public removeCourse(id: string): void {
-    courses.splice(courses.findIndex(course => course.id === +id), 1)
-    console.log(courses)
+    console.log(this.coursesMock)
+    this.coursesMock.splice(this.coursesMock.findIndex(course => course.id == +id), 1)
```
Comment (AndriyLeskiv) [inline src/app/services/courses.service.ts:41]: try use filter  this.coursesMock =  this.coursesMock.filter((course) => {
      return course.id !== +id;
    })

File: src/app/guards/not-auth/not-auth.guard.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import { Injectable} from '@angular/core';
+import {ActivatedRouteSnapshot, CanActivate, RouterStateSnapshot, UrlTree} from '@angular/router';
+import {Observable} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class NotAuthGuard implements CanActivate {
+
+  public isAuthenticated(): boolean {
```
Comment (AndriyLeskiv) [inline src/app/guards/not-auth/not-auth.guard.ts:10]: use AuthService.isAuthenticated()

File: src/app/guards/not-auth/not-auth.guard.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    const email = localStorage.getItem('email');
+    const password = localStorage.getItem('password');
+    return !!(email && password);
+  }
+
+  canActivate(
+    route: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
```
Comment (AndriyLeskiv) [inline src/app/guards/not-auth/not-auth.guard.ts:18]: remove unneeded parameters

File: src/app/modules/courses/components/edit-course/edit-course.component.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+  public editCurrentCourse(): void {
+    this.coursesService.updateCourse(this.course.id, this.course)
+  }
+
+  ngOnInit(): void {
+  }
+
+  ngOnDestroy() {
+    this.activatedRoute.params.subscribe().unsubscribe()
```
Comment (AndriyLeskiv) [inline src/app/modules/courses/components/edit-course/edit-course.component.ts:38]: what is this??? https://blog.bitsrc.io/6-ways-to-unsubscribe-from-observables-in-angular-ab912819a78f

File: src/app/modules/courses/components/edit-course/edit-course.component.ts [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+})
+
+export class EditCourseComponent implements OnInit, OnDestroy {
+  public course: ICourse;
+
+  public save: string = 'Save';
+  public cancel: string = 'Cancel';
+
+  public notFoundText = '404 error';
+  public text = "sorry, your page doesn't exists";
```
Comment (AndriyLeskiv) [inline src/app/modules/courses/components/edit-course/edit-course.component.ts:20]: use router.navigate to 404 if course is absent

File: src/app/services/courses.service.ts (hunk 14 above)
Comment (yarikkk13) [reply src/app/services/courses.service.ts:41]: done

File: src/app/guards/not-auth/not-auth.guard.ts (hunk 15 above)
Comment (yarikkk13) [reply src/app/guards/not-auth/not-auth.guard.ts:10]: done

File: src/app/guards/not-auth/not-auth.guard.ts (hunk 16 above)
Comment (yarikkk13) [reply src/app/guards/not-auth/not-auth.guard.ts:18]: done

File: src/app/modules/courses/components/edit-course/edit-course.component.ts (hunk 17 above)
Comment (yarikkk13) [reply src/app/modules/courses/components/edit-course/edit-course.component.ts:38]: done

**Pull request merged**